edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs

```diff
@@ -27,23 +31,54 @@ public Task StoreMetricsAsync(IEnumerable<Metric> metrics)
 
         public async Task<IEnumerable<Metric>> GetAllMetricsAsync()
         {
-            return (await this.dataStore.GetBatch(0, 1000))
-                    .SelectMany(((long offset, IEnumerable<Metric> metrics) storeResult) =>
+            // Note: async/await is needed b/c ToListAsync returns ValueTask instead of Task.
+            return await this.GetAllMetricsInternal().ToListAsync();
+        }
+
+        async IAsyncEnumerable<Metric> GetAllMetricsInternal([EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            HashSet<long> maxAgeEntriesToRemove = new HashSet<long>();
+
+            foreach ((long offset, IEnumerable<Metric> metrics) in await this.dataStore.GetBatch(0, 1000, cancellationToken))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                // This makes the assumption that metrics are stored in batches with matching timestamps.
+                // This is the case b/c the metrics are stored immediately after being scraped.
+                if (metrics.First().TimeGeneratedUtc < DateTime.UtcNow - this.maxMetricAge)
+                {
+                    // TODO: log removed
```

Comment:
Can we add the logging in this PR? I imagine we'd count up the total # of removed entries and report it in a single log statement (at least # removed entries, although time range would be nice to have)

Reply:
added logging. I didn't bother logging at the end. The only time multiple scrapes will be deleted at once is if the device was powered off for an extended period of time.

Reply:
Looks good.